source/log_formatter.c

```diff
@@ -0,0 +1,199 @@
+     */
+    int total_length = required_length + MAX_LOG_LINE_PREFIX_SIZE;
+    struct aws_string *raw_string = (struct aws_string *)aws_mem_acquire(formatter->allocator, sizeof(struct aws_string) + total_length);
+    if (raw_string == NULL) {
+        goto cleanup;
+    }
+
+    char *log_line_buffer = (char *)raw_string->bytes;
+    size_t current_index = 0;
+
+    /*
+     * Begin the log line with "[<Log Level>] ["
+     */
+    const char *level_string = NULL;
+    if (aws_logging_log_level_to_string(level, &level_string)) {
+        goto cleanup;
+    }
+
+    int log_level_length = snprintf(log_line_buffer, total_length, "[%s] [", level_string);
+    if (log_level_length < 0) {
+        goto cleanup;
+    }
+
+    current_index += log_level_length;
+
+    /*
+     * Add the timestamp.  To avoid copies and allocations, do some byte buffer tomfoolery.
+     *
+     * First, make a byte_buf that points to the current position in the output string
+     */
+    struct aws_byte_buf timestamp_buffer;
+    timestamp_buffer.allocator = formatter->allocator;
+    timestamp_buffer.buffer = (uint8_t *)log_line_buffer + current_index;
+    timestamp_buffer.capacity = total_length - current_index;
+    timestamp_buffer.len = 0;
+
+    /*
+     * Output the current time to the byte_buf
+     */
+    struct aws_date_time current_time;
+    aws_date_time_init_now(&current_time);
+
+    result = aws_date_time_to_utc_time_str(&current_time, impl->date_format, &timestamp_buffer);
+    if (result != AWS_OP_SUCCESS) {
+        goto cleanup;
+    }
+
+    current_index += timestamp_buffer.len;
+
+    /*
+     * Add thread id and user content separator (" - ")
+     */
+    uint64_t current_thread_id = aws_thread_current_thread_id();
+    int thread_id_written = snprintf(log_line_buffer + current_index, total_length - current_index, "] [%"PRIu64"] - ", current_thread_id);
+    if (thread_id_written < 0) {
+        goto cleanup;
+    }
+
+    current_index += thread_id_written;
+
+    /*
+     * Now write the actual data requested by the user
+     */
+#ifdef WIN32
+    int written_count = vsnprintf_s(log_line_buffer + current_index, total_length - current_index, _TRUNCATE, format, args);
+#else
+    int written_count = vsnprintf(log_line_buffer + current_index, total_length - current_index, format, args);
+#endif // WIN32
+    if (written_count < 0) {
+        goto cleanup;
+    }
+
+    /*
+     * End with a newline.
+     */
+    current_index += written_count;
+    written_count = snprintf(log_line_buffer + current_index, total_length - current_index, "\n");
+    if (written_count < 0) {
+        goto cleanup;
+    }
+
+    result = AWS_OP_SUCCESS;
+
+    *(struct aws_allocator **) (&raw_string->allocator) = formatter->allocator;
+    *(size_t *) (&raw_string->len) = current_index + written_count;
+
+    *formatted_output = raw_string;
+
```

Comment:
Could just return AWS_OP_SUCCESS here and not have to worry about result at all. Would remove the result check in cleanup:

Reply:
Will do